protocol-units/settlement/mcr/contracts/test/MCR.t.sol

```diff
@@ -0,0 +1,340 @@
+        vm.prank(signer1);
+        mcr.stakeGenesis{value : 34 ether}();
+        vm.prank(signer2);
+        mcr.stakeGenesis{value : 33 ether}();
+        vm.prank(signer3);
+        mcr.stakeGenesis{value : 33 ether}();
+
+        // signer3 will be dishonest
+        MCR.BlockCommitment memory dishonestCommitment = MCR.BlockCommitment({
+            height : 1,
+            commitment: keccak256(abi.encodePacked(uint256(3), uint256(2), uint256(1))),
+            blockId: keccak256(abi.encodePacked(uint256(3), uint256(2), uint256(1)))
+        });
+        vm.prank(signer3);
+        mcr.submitBlockCommitment(dishonestCommitment);
+
+        // signer3 will try to sign again
+        vm.prank(signer3);
+        vm.expectRevert("Validator has already committed to a block at this height");
+        mcr.submitBlockCommitment(dishonestCommitment);
+
+        // signer1 and signer2 will be honest
+        MCR.BlockCommitment memory honestCommitment = MCR.BlockCommitment({
+            height : 1,
+            commitment: keccak256(abi.encodePacked(uint256(1), uint256(2), uint256(3))),
+            blockId: keccak256(abi.encodePacked(uint256(1), uint256(2), uint256(3)))
+        });
+        vm.prank(signer1);
+        mcr.submitBlockCommitment(honestCommitment);
+        vm.prank(signer2);
+        mcr.submitBlockCommitment(honestCommitment);
+
+        // the honest block should be accepted
+        assert(mcr.getAcceptedCommitmentAtBlockHeight(1).commitment == honestCommitment.commitment);
+        assert(mcr.getAcceptedCommitmentAtBlockHeight(1).blockId == honestCommitment.blockId);
+        assert(mcr.getAcceptedCommitmentAtBlockHeight(1).height == 1);
+
+    }
+
+    address[] honestSigners = new address[](0);
+    address[] dishonestSigners = new address[](0);
+
+    function testChangingValidatorSet() public {
+
+        vm.pauseGasMetering();
+
+        uint256 blockTime = 5;
+        MCR mcr = new MCR(
+            5, // 5 second block time 
+            128,
+            100 ether, // should accumulate 100 ether
+            0
+        );
+
+        vm.warp(blockTime);
+
+        // three well-funded signers
+        address payable signer1 = payable(vm.addr(1)); 
+        vm.deal(signer1, 100 ether);
+        address payable signer2 = payable(vm.addr(2));
+        vm.deal(signer2, 100 ether);
+        address payable signer3 = payable(vm.addr(3));
+        vm.deal(signer3, 100 ether);
+
+        // have them participate in the genesis ceremony
+        vm.prank(signer1);
+        mcr.stakeGenesis{value : 34 ether}();
+        vm.prank(signer2);
+        mcr.stakeGenesis{value : 33 ether}();
+        vm.prank(signer3);
+        mcr.stakeGenesis{value : 33 ether}();
+
+        // honest signers
+        honestSigners.push(signer1);
+        honestSigners.push(signer2);
+
+        // dishonest signers
+        dishonestSigners.push(signer3);
+
+        for (uint i = 0; i < 50; i++) {
```

Comment:
Can we `uint256 epochs = 50`, then
`for (uint i = 0; i < epochs; i++) {`

Without reading the readme, its not obvious to the reader  what the 50 represents.

Reply:
It's not exactly epochs. We have have ticks of +1 vm time in the inner loop. I can call it reorgs, because there is one validator set reorg requested per iteration.